# Churn Analysis of a Telecom Company

# About Dataset

**Content:**<br>
The **Telco Customer Churn** dataset includes a variety of variables or features that provide information about customers and their interactions with a telecommunications company, each row represents a customer, each column contains customer’s attributes.
1. **Customer ID**: A unique identifier for each customer.
2. **Gender**: The gender of the customer (e.g., Male, Female).
3. **Senior Citizen**: Whether the customer is a senior citizen or not, 1: yes / 0: no.
4. **Partner**: Whether the customer has a partner (Yes/No).
5. **Dependents**: Whether the customer has dependents (children, relatives...), (Yes/No).
6. **Tenure**: The number of months the customer has stayed with the company.
7. **Phone Service**: Whether the customer has phone service provided by the company (Yes/No).
8. **Multiple Lines**: Whether the customer has multiple lines (e.g., Yes, No, No phone service).
9. **Internet Service**: Type of internet service subscribed (e.g., DSL, Fiber optic, No).
10. **Online Security**: Whether the customer has online security service (e.g., Yes, No, No internet service).
11. **Online Backup**: Whether the customer has online backup service (e.g., Yes, No, No internet service).
12. **Device Protection**: Whether the customer has device protection service (e.g., Yes, No, No internet service).
13. **Tech Support**: Whether the customer has tech support service (e.g., Yes, No, No internet service).
14. **Streaming TV**: Whether the customer has streaming TV service (e.g., Yes, No, No internet service).
15. **Streaming Movies**: Whether the customer has streaming movie service (e.g., Yes, No, No internet service).
16. **Contract**: The type of contract the customer has (e.g., Month-to-month, One year, Two year).
17. **Paperless Billing**: Whether the customer has opted for paperless billing (Yes/No).
18. **Payment Method**: The method of payment used by the customer (e.g., Electronic check, Credit card, Bank transfer, Mailed check).
19. **Monthly Charges**: The amount charged to the customer on a monthly basis.
20. **Total Charges**: The total amount charged to the customer over the entire tenure.
21. **Churn**: Whether the customer churned (cancelled the service) or not (Yes/No).

**Objective**
- Our target variable is **Churn**, the objective is to analyze what factors affect the churn rate in the company.

# 1. Importing Data from Python and Reading Dataset

In [356]:
import pandas as pd # for data processing
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')

# 2. Inspecting the data

In [357]:
# Inspecting the first few rows 
df.head(3)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   

      PaymentMethod MonthlyCharges  TotalCharges Churn  
0  Electronic check          29.85         29.85    No  
1      Mailed check          56.95        1889.5    No  
2      Mailed check          53.85        108.15   Yes  

[3 rows x 21 columns]

In [358]:
# Inspecting the last few rows
df.tail(3)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
7040    No  
7041   Yes  
7042    No  

[3 rows x 21 columns]

In [359]:
df.shape
#(rows,columns)

(7043, 21)

In [360]:
# checking if customer ID's are unique
len(list(df['customerID'].unique()))

7043

In [361]:
# Looking at info of each column such as column name, number of rows, number of non_null values and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


The variable TotalCharges is a numeric variable, but it is of type 'object' in our data set, we need to change it to float.

In [362]:
# Description of some numerical variables
df.describe().round(3)

SeniorCitizen    tenure  MonthlyCharges
count       7043.000  7043.000        7043.000
mean           0.162    32.371          64.762
std            0.369    24.559          30.090
min            0.000     0.000          18.250
25%            0.000     9.000          35.500
50%            0.000    29.000          70.350
75%            0.000    55.000          89.850
max            1.000    72.000         118.750

- 16% of customers are senior citizens.
- Customers have stayed 32 months on average, that's 2.5+ years.
- Nothing fishy about our data...

In [363]:
# Description of categorical variables
df.describe(include = 'object').T

count unique               top  freq
customerID        7043   7043        7590-VHVEG     1
gender            7043      2              Male  3555
Partner           7043      2                No  3641
Dependents        7043      2                No  4933
PhoneService      7043      2               Yes  6361
MultipleLines     7043      3                No  3390
InternetService   7043      3       Fiber optic  3096
OnlineSecurity    7043      3                No  3498
OnlineBackup      7043      3                No  3088
DeviceProtection  7043      3                No  3095
TechSupport       7043      3                No  3473
StreamingTV       7043      3                No  2810
StreamingMovies   7043      3                No  2785
Contract          7043      3    Month-to-month  3875
PaperlessBilling  7043      2               Yes  4171
PaymentMethod     7043      4  Electronic check  2365
TotalCharges      7043   6531                      11
Churn             7043      2                No  5174

# 3. Data Wrangling

### Changing data type

In [364]:
# Converting this variable to numeric
# df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])
# Get an error: we have empty strings in this column.

In [365]:
# We have 11 instances with an empty string in the TotalCharges columns
df.loc[df['TotalCharges'] == ' ', :]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
488            No  No phone service             DSL                  Yes  ...   
753           Yes                No              No  No internet service  ...   
936           Yes                No             DSL                  Yes  ...   
1082          Yes               Yes              No  No internet service  ...   
1340           No  No phone service             DSL                  Yes  ...   
3331          Yes                No              No  No internet service  ...   
3826          Yes               Yes              No  No internet service  ...   
4380          Yes                No              No  No internet service  ...   
5218          Yes                No              No  No internet service  ...   
6670          Yes               Yes             DSL                   No  ...   
6754          Yes               Yes             DSL                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
488                   Yes                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                   No                  Yes   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                   No                  Yes                   No   

          StreamingMovies  Contract PaperlessBilling  \
488                    No  Two year              Yes   
753   No internet service  Two year               No   
936                   Yes  Two year               No   
1082  No internet service  Two year               No   
1340                   No  Two year               No   
3331  No internet service  Two year               No   
3826  No internet service  Two year               No   
4380  No internet service  Two year               No   
5218  No internet service  One year              Yes   
6670                   No  Two year               No   
6754                   No  Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)          52.55                  No  
753                Mailed check          20.25                  No  
936                Mailed check          80.85                  No  
1082               Mailed check          25.75                  No  
1340    Credit card (automatic)          56.05                  No  
3331               Mailed check          19.85                  No  
3826               Mailed check          25.35                  No  
4380               Mailed check          20.00                  No  
5218               Mailed check          19.7

In [366]:
# Finding a relation between monthly and total charges with tenure
df[['tenure', 'MonthlyCharges', 'TotalCharges']].head()

tenure  MonthlyCharges TotalCharges
0       1           29.85        29.85
1      34           56.95       1889.5
2       2           53.85       108.15
3      45           42.30      1840.75
4       2           70.70       151.65

It seems like ***Totalcharges ≈ tenure * MonthlyCharges*** but this is not accurate, there could be discounts or extra costs for delays in payment.

All of the customers who have not stayed with the company for any period of time have a **Tenure** value of 0. We can replace the **TotalCharges** values with 0, based on our previous discovery. These two variables are positively correlated.

In [367]:
# Replacing string values with 0
df['TotalCharges'].replace(' ',0,inplace=True)

In [368]:
# We can cast it into a float now
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])
df['TotalCharges'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7043 entries, 0 to 7042
Series name: TotalCharges
Non-Null Count  Dtype  
--------------  -----  
7043 non-null   float64
dtypes: float64(1)
memory usage: 55.1 KB


In [369]:
# Confirming the correlation
df[['tenure', 'MonthlyCharges','TotalCharges']].corr()

tenure  MonthlyCharges  TotalCharges
tenure          1.000000        0.247900      0.826178
MonthlyCharges  0.247900        1.000000      0.651174
TotalCharges    0.826178        0.651174      1.000000

### Binning

We can bin 3 variables in out data:
1. tenure
2. MonthlyCharges
3. TotalCharges

In [401]:
# Inspecting our variables:
df[['tenure','MonthlyCharges','TotalCharges']].agg(['mean','min','max','count']).round(2)


tenure  MonthlyCharges  TotalCharges
mean     32.37           64.76       2279.73
min       0.00           18.25          0.00
max      72.00          118.75       8684.80
count  7043.00         7043.00       7043.00

In [405]:
# Binning will make our analysis easier
bins_tenure = [0,32,64,96]
bins_monthly = [0,20,64,128]
bins_total = [0,2000,4000,9000]
df['Tenure_bin'] = pd.cut(df['tenure'], bins=bins_tenure, labels=['00-32', '32-64', '64-119'])
df['MonthlyCharges_bin'] = pd.cut(df['MonthlyCharges'], bins=bins_monthly, labels=['Low', 'Medium', 'High'])
df['TotalCharges_bin'] = pd.cut(df['TotalCharges'], bins=bins_total, labels=['Low', 'Medium', 'High'])
df[['Tenure_bin','MonthlyCharges_bin', 'TotalCharges_bin']].head()

Tenure_bin MonthlyCharges_bin TotalCharges_bin
0      00-32             Medium              Low
1      32-64             Medium              Low
2      00-32             Medium              Low
3      32-64             Medium              Low
4      00-32               High              Low

# 4. Data Analysis

## What are the factors (variables) affecting customer churn?

In [370]:
# Checking all variable names
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

### Is customer churn gender neutral?

In [371]:
df[['gender', 'Churn']].groupby('gender').value_counts().reset_index()

gender Churn  count
0  Female    No   2549
1  Female   Yes    939
2    Male    No   2625
3    Male   Yes    930

Gender does not affect churn, almost same number or customers left the company in both gender groups.

### Do senior citizens churn more often than others?

In [372]:
pd.DataFrame(df[['Churn','SeniorCitizen']].groupby('SeniorCitizen').value_counts(normalize = True).mul(100)).rename(columns = {'proportion':'percentage'}).round(2)

percentage
SeniorCitizen Churn            
0             No          76.39
              Yes         23.61
1             No          58.32
              Yes         41.68

A higher percentage of senior citizens churned (41.6%) compared to non-senior citizens (23.6%). This suggests that being a senior citizen may be associated with churning, but further analysis must be done to understand the reasons behind these patterns.

### Does having a partner affect a customer's behavior, what about dependents?

In [373]:
pd.DataFrame(df[['Churn','Partner']].groupby('Partner').value_counts(normalize = True).mul(100).round(2)).rename(columns = {'proportion':'percentage'})

percentage
Partner Churn            
No      No          67.04
        Yes         32.96
Yes     No          80.34
        Yes         19.66

In [374]:
pd.DataFrame(df[['Churn','Dependents']].groupby('Dependents').value_counts(normalize = True).mul(100).round(2)).rename(columns = {'proportion':'percentage'})

percentage
Dependents Churn            
No         No          68.72
           Yes         31.28
Yes        No          84.55
           Yes         15.45

- Customers who have partners and dependents are less likely to churn

### Do 'PhoneService' and'MultipleLines' play a significant role?

In [375]:
pd.DataFrame(df[['PhoneService', 'MultipleLines', 'Churn']].groupby('PhoneService').value_counts(normalize=True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
PhoneService MultipleLines    Churn            
No           No phone service No      75.073314
                              Yes     24.926686
Yes          No               No      39.946549
             Yes              No      33.343814
                              Yes     13.362679
             No               Yes     13.346958

Having a phone service and multiple lines do not seem to significantly affect the churn rate.

### Does subscribing to an Internet service affect churn? What about other related services?

In [376]:
cols = ['InternetService', 'Churn']
pd.DataFrame(df[cols].groupby('InternetService').value_counts(normalize=True)).mul(100).rename(columns={'proportion':'Percentage'})

Percentage
InternetService Churn            
DSL             No      81.040892
                Yes     18.959108
Fiber optic     No      58.107235
                Yes     41.892765
No              No      92.595020
                Yes      7.404980

- Most of customers who have have not subscribed to Internet Service didn't churn.
- Of those who do have Internet service, customers with a DSL provider seem content and did not churn, while almost 42% of those with a Fiber optic provider churned.


In [377]:
pd.DataFrame(df[cols].groupby('Churn').value_counts(normalize=True)).mul(100).rename(columns={'proportion':'Percentage'})

Percentage
Churn InternetService            
No    DSL               37.920371
      Fiber optic       34.770004
      No                27.309625
Yes   Fiber optic       69.395399
      DSL               24.558587
      No                 6.046014

Of those who churned, almost twice the percentage of customers left the company.

In [378]:
# Why would customers with a Fiber optic provider churn?
df[['InternetService' ,'MonthlyCharges']].groupby('InternetService').mean().round(2).reset_index()

InternetService  MonthlyCharges
0             DSL           58.10
1     Fiber optic           91.50
2              No           21.08

It appears that the high monthly cost of Fiber optic Internet service could be why some customers who subscribed are churning.

#### Online Security

In [379]:
pd.DataFrame(df[['OnlineSecurity','Churn']].groupby('OnlineSecurity').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
OnlineSecurity      Churn            
No                  No      58.233276
                    Yes     41.766724
No internet service No      92.595020
                    Yes      7.404980
Yes                 No      85.388806
                    Yes     14.611194

In [380]:
pd.DataFrame(df[['OnlineSecurity','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn OnlineSecurity                 
No    No                    39.369927
      Yes                   33.320448
      No internet service   27.309625
Yes   No                    78.170144
      Yes                   15.783842
      No internet service    6.046014

#### Online Backup

In [381]:
pd.DataFrame(df[['OnlineBackup','Churn']].groupby('OnlineBackup').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
OnlineBackup        Churn            
No                  No      60.071244
                    Yes     39.928756
No internet service No      92.595020
                    Yes      7.404980
Yes                 No      78.468506
                    Yes     21.531494

#### Device Protection

In [382]:
pd.DataFrame(df[['DeviceProtection','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn DeviceProtection               
No    No                    36.412833
      Yes                   36.277542
      No internet service   27.309625
Yes   No                    64.794007
      Yes                   29.159979
      No internet service    6.046014

#### Tech Support

In [383]:
pd.DataFrame(df[['TechSupport','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn TechSupport                    
No    No                    39.176652
      Yes                   33.513722
      No internet service   27.309625
Yes   No                    77.367576
      Yes                   16.586410
      No internet service    6.046014

#### Streaming TV

In [384]:
pd.DataFrame(df[['StreamingTV','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn StreamingTV                    
No    Yes                   36.586780
      No                    36.103595
      No internet service   27.309625
Yes   No                    50.401284
      Yes                   43.552702
      No internet service    6.046014

#### Streaming Movies

In [385]:
pd.DataFrame(df[['StreamingMovies','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn StreamingMovies                
No    Yes                   36.992656
      No                    35.697719
      No internet service   27.309625
Yes   No                    50.187266
      Yes                   43.766720
      No internet service    6.046014

In [386]:
pd.DataFrame(df[['StreamingMovies','Churn']].groupby('StreamingMovies').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
StreamingMovies     Churn            
No                  No      66.319569
                    Yes     33.680431
No internet service No      92.595020
                    Yes      7.404980
Yes                 No      70.058565
                    Yes     29.941435

- Customers who use Online Security, Online Backup, Device Protection, and Tech Support services are less likely to churn, but not significantly so.
- Streaming TV and Streaming Movies services do not have much of an influence on customer curn.

### Are contract types related to customer churn?

In [387]:
pd.DataFrame(df[['Contract','Churn']].groupby('Contract').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Contract       Churn            
Month-to-month No      57.290323
               Yes     42.709677
One year       No      88.730482
               Yes     11.269518
Two year       No      97.168142
               Yes      2.831858

- Most people who signed one-year and two-year contracts didn't leave the company.
- Let's look closer at the month-to-month contracts.

In [388]:
pd.DataFrame(df[['Contract','Churn']].groupby('Churn').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
Churn Contract                  
No    Month-to-month   42.906842
      Two year         31.832238
      One year         25.260920
Yes   Month-to-month   88.550027
      One year          8.881755
      Two year          2.568218

- It's significant that 88.5% of people who churned have signed month-to-month contracts.

**The contract does indeed play a significant role in customer churn.**

### Do customers with paperless billing churn more often?

In [389]:
pd.DataFrame(df[['PaperlessBilling', 'Churn']].groupby('PaperlessBilling').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
PaperlessBilling Churn            
No               No      83.669916
                 Yes     16.330084
Yes              No      66.434908
                 Yes     33.565092

- The majority of customers with and without paperless billing do not churn, but of those who do leave the company, they're twice as likely to leave when billed with paperless billing.

### Do certain payment methods increase churn? Which payment method makes customers churn more often?

In [406]:
pd.DataFrame(df[['PaymentMethod', 'Churn']].groupby('PaymentMethod').value_counts(normalize = True).mul(100)).rename(columns={'proportion':'Percentage'})

Percentage
PaymentMethod             Churn            
Bank transfer (automatic) No      83.290155
                          Yes     16.709845
Credit card (automatic)   No      84.756899
                          Yes     15.243101
Electronic check          No      54.714588
                          Yes     45.285412
Mailed check              No      80.893300
                          Yes     19.106700

Customers who pay using electronic checks are more than twice likely to leave the company compared to those who pay with other methods.